server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -810,8 +810,10 @@ private static void parseNonDynamicArray(
             boolean objectWithFallbackSyntheticSource = false;
             if (mapper instanceof ObjectMapper objectMapper) {
                 mode = getSourceKeepMode(context, objectMapper.sourceKeepMode());
-                objectWithFallbackSyntheticSource = (mode == Mapper.SourceKeepMode.ALL
-                    || (mode == Mapper.SourceKeepMode.ARRAYS && objectMapper instanceof NestedObjectMapper == false));
+                objectWithFallbackSyntheticSource = mode == Mapper.SourceKeepMode.ALL
+                    || (mode == Mapper.SourceKeepMode.ARRAYS && objectMapper instanceof NestedObjectMapper == false)
+                    // Inside nested objects we always store object arrays as a workaround for #115261.
+                    || context.inNestedScope();
```

Comment:
I think it's more nuanced.. the check above bypasses storing array source for nested objects, as they exist for this exact reason.

Can we detect that there are deeply nested arrays and store source only for them? Maybe 
`|| (context.inNestedScope() && objectMapper instanceof NestedObjectMapper == false))`

Reply:
That should work!